File: test-cases/check_output/not_supported/eval.js

var a = 3, b = 4;

function f1() {
    var a = 1;
    eval('var b = 2; function g() { console.log("g"); }');
    console.log(a, b);
    g();
}
f1();
console.log(a, b); // 1, 1


// 测试 eval 中函数的作用域
// g, f 都有效 
function f2() {
    eval('function g() { console.log("g"); }');
    eval('{ function f() { console.log("f"); } }');

    g();
    f();
}
f2();


function f() {
    {
        let a = 1;
        eval('function g() { console.log("g"); }');

        console.log(a);
        {
            let x = 2;
            console.log(a);

            g();
        }
    }
}
f();

File: api-built-in/eval.cpp
﻿//
//  Error.cpp
//  TinyJS
//
//  Created by henry_xiao on 2022/7/31.
//

#include "BuiltIn.hpp"


static JsValue __errorPrototype;
static JsValue __errorToStringPrefix;
static StringView SS_ERR_STRING_PREFIX = makeCommonString("Error: ");


// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval
static void _eval(VMContext *ctx, const JsValue &thiz, const Arguments &args) {
    auto runtime = ctx->runtime;

    if (args.count == 0) {
        ctx->stack.push_back(jsValueUndefined);
        return;
    }

    auto v = args[0];
    if (v.type == JDT_STRING) {
        VecVMStackScopes stackScopesGlobal;

        auto code = runtime->getString(v);
        auto stackScopes = ctx->stackScopesForNativeFunctionCall;
        if (stackScopes == nullptr) {
            stackScopesGlobal.push_back(runtime->globalScope());
            stackScopes = &stackScopesGlobal;
        }

        ctx->vm->eval((cstr_t)code.utf8Str().data, code.utf8Str().len, ctx, *stackScopes, args);
    } else if (v.type == JDT_CHAR) {
        ctx->retValue = jsValueUndefined;
    } else {
        ctx->retValue = v;
    }
}

static JsLibProperty errorFunctions[] = {
    { "name", nullptr, "eval" },
    { "length", nullptr, nullptr, jsValueLength1Property },
};

void registerEval(VMRuntimeCommon *rt) {
    setGlobalLibObject("eval", rt, errorFunctions, CountOf(errorFunctions), _eval, jsValuePrototypeFunction);
}
